<a href="https://colab.research.google.com/github/Hareg2024/User-churn/blob/main/waze.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Connect to bigquery

In [5]:
#import libraries
from google.cloud import bigquery
from google.colab import auth

#authonticate
auth.authenticate_user()

#Initialize the client from bigquery
PROJECT_id = 'userchurn-431116'
client =bigquery.Client(project= PROJECT_id, location='US')


In [16]:
#get the dataset and table , one table at a time
dataset_ref = client.dataset('wazedata', project='userchurn-431116')
table_ref = dataset_ref.table('rawdata')
table = client.get_table(table_ref)
table.schema

[SchemaField('ID', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('label', 'STRING', 'NULLABLE', None, None, (), None),
 SchemaField('sessions', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('drives', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('total_sessions', 'FLOAT', 'NULLABLE', None, None, (), None),
 SchemaField('n_days_after_onboarding', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('total_navigations_fav1', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('total_navigations_fav2', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('driven_km_drives', 'FLOAT', 'NULLABLE', None, None, (), None),
 SchemaField('duration_minutes_drives', 'FLOAT', 'NULLABLE', None, None, (), None),
 SchemaField('activity_days', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('driving_days', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('device', 'STRING', 'NULLABLE', None, None, (), None)]

In [12]:
# bring the second table for the test table
table_ref = dataset_ref.table('testdata')
testtable = client.get_table(table_ref)
testtable.schema

[SchemaField('ID', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('label', 'STRING', 'NULLABLE', None, None, (), None),
 SchemaField('sessions', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('drives', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('total_sessions', 'FLOAT', 'NULLABLE', None, None, (), None),
 SchemaField('n_days_after_onboarding', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('total_navigations_fav1', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('total_navigations_fav2', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('driven_km_drives', 'FLOAT', 'NULLABLE', None, None, (), None),
 SchemaField('duration_minutes_drives', 'FLOAT', 'NULLABLE', None, None, (), None),
 SchemaField('activity_days', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('driving_days', 'INTEGER', 'NULLABLE', None, None, (), None),
 SchemaField('device', 'STRING', 'NULLABLE', None, None, (), None),
 SchemaField('typ

In [51]:
# convert the 2 tables into dataframe
df = client.list_rows(table=table).to_dataframe()
df.head()


ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   1806  retained         0       0       37.168690                     3098   
1   3451  retained         0       0      210.090088                     2873   
2   6375  retained         0       0       99.767601                     1386   
3   8122  retained         0       0      132.322350                       89   
4  10586   churned         0       0      222.107433                      980   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     236                       0       5570.745476   
1                     210                      35      12012.320180   
2                      21                       0       4169.721686   
3                     174                     104       5044.893956   
4                       0                       6       4932.029466   

   duration_minutes_drives  activity_days  driving_days   device  
0               943.950161              6             0  Android  
1              4804.667454              1             0  Android  
2              3154.856024              2             0  Android  
3              2205.094060              4             0   iPhone  
4              2656.000047              3             0  Android

In [33]:
# the second table is a view table and cannot be directly converted to dataframe , we are gonna query the view table
sql = """
SELECT *
FROM `userchurn-431116.wazedata.testdata`
"""

query_job = client.query(sql)
results = query_job.result()  # Waits for the query to finish
df_test = results.to_dataframe()
df_test.head()

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   1806  retained         0       0       37.168690                     3098   
1   3451  retained         0       0      210.090088                     2873   
2   6375  retained         0       0       99.767601                     1386   
3   8122  retained         0       0      132.322350                       89   
4  10586   churned         0       0      222.107433                      980   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     236                       0       5570.745476   
1                     210                      35      12012.320180   
2                      21                       0       4169.721686   
3                     174                     104       5044.893956   
4                       0                       6       4932.029466   

   duration_minutes_drives  activity_days  driving_days   device  type  
0               943.950161              6             0  Android  test  
1              4804.667454              1             0  Android  test  
2              3154.856024              2             0  Android  test  
3              2205.094060              4             0   iPhone  test  
4              2656.000047              3             0  Android  test

# Build Model

## Install pycaret to access auto ml

In [23]:
!pip install pycaret

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.4/60.4 kB 4.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 163.8/163.8 kB 11.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.4/50.4 kB 3.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 486.1/486.1 kB 22.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 106.8/106.8 kB 8.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.8/21.8 MB 47.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.9/81.9 kB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.2/302.2 kB 20.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.9/79.9 MB 7.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 80.7/80.7 kB 6.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 65.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.1/12.1 MB 94.0 MB/s eta 0:00:00
   ━━

# Code and Train

In [27]:
#get the classification model: user chaurn is catagorical y
from pycaret.classification  import *

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  Int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  Int64  
 3   drives                   14999 non-null  Int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  Int64  
 6   total_navigations_fav1   14999 non-null  Int64  
 7   total_navigations_fav2   14999 non-null  Int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  Int64  
 11  driving_days             14999 non-null  Int64  
 12  device                   14999 non-null  object 
dtypes: Int64(8), float64(3), object(2)
memory usage: 1.6+ MB


In [52]:
#handle missing values and duplicates
df=df.dropna(axis=0)

#df=-df.drop_duplicates()

df.duplicated().sum()

0

In [53]:
# setup model
setup(data=df,
      target='label',
      session_id=123,
      ignore_features=['ID','type'],
      categorical_features=['device'])

In [54]:
#compare all classification model with sklearn models based on evaluation metrics
compare_models()

Processing:   0%|          | 0/65 [00:00<?, ?it/s]

GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='log_loss', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_samples_leaf=1,
                           min_samples_split=2, min_weight_fraction_leaf=0.0,
                           n_estimators=100, n_iter_no_change=None,
                           random_state=123, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

### The GradientBoosting model is selected based on evaluation metrics,
* Accuracy= 0.82,Recall(how many of the actual positives were correctly identified)=0.82  and
* Precission (the accuracy of positive predictions)=0.78

In [55]:
# use the best model
gbst = create_model('gbc')

Processing:   0%|          | 0/4 [00:00<?, ?it/s]

In [56]:
# use the model to predict on holdout data
gbst_pred = predict_model(gbst)

In [58]:
df_final=gbst_pred.copy()
df_final.head(20)

sessions  drives  total_sessions  n_days_after_onboarding  \
1487        126     110      164.833176                     2486   
6682          9       9      440.216736                     1638   
7663        155     127      286.282410                      782   
12677        68      55      119.666885                      957   
3436         34      28      136.838440                      841   
234          21      17       21.705410                      955   
13280       131     131      170.967514                     3185   
5637         30      25       36.760063                     3436   
7278         15      15       29.705603                       85   
12796       129     104      131.545685                     1755   
3190        172     142      252.687897                     1823   
2895         27      22      224.553787                     3456   
3949         26      22      137.166016                       82   
14488        31      27      226.025604                      821   
2989         51      42      102.138802                      395   
7138        154     123      240.163834                     2353   
2497         74      62      144.158218                     2409   
11283        12      11       54.689850                     2964   
9569         30      29       55.490204                      823   
450          46      37      259.187286                     3369   

       total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
1487                      432                       0       2159.167725   
6682                      139                      32       1877.869507   
7663                      721                       4       6827.522949   
12677                      34                      25       1369.640503   
3436                      393                      10       3064.986328   
234                       237                       0       4941.263184   
13280                     191                     191       4903.722656   
5637                      157                      69       3589.366455   
7278                       61                      41        629.386047   
12796                     119                       5       9598.928711   
3190                        0                      28       4468.734863   
2895                      368                       0       5911.304688   
3949                      437                      28       2167.763428   
14488                      18                       0       2040.500366   
2989                       45                      81       1115.136719   
7138                      184                      48       2663.476074   
2497                       29                       2       1974.607300   
11283                     224                       0       1224.017822   
9569                       12                      66       3392.948975   
450                         0                       0       3664.357178   

       duration_minutes_drives  activity_days  driving_days   device  \
1487                501.427673              6             1   iPhone   
6682               1016.116577             18            11   iPhone   
7663               3161.975342             12            12  Android   
12677               402.879852             27            21  Android   
3436               2048.386719              9             5   iPhone   
234                4039.858154              2             0   iPhone   
13280              3332.038574             22            22   iPhone   
5637               1680.339111             16             9   iPhone   
7278                476.107086             18            12   iPhone   
12796              2376.764648             29            21  Android   
3190               2400.274414              4             4   iPhone   
2895               2665.157959              4             4   iPhone   
3949               1735.421997              6             6  Andro

In [63]:
churned= df_final['label']
churned.value_counts()


label
retained    3529
churned      761
Name: count, dtype: int64

In [64]:
churned= df_final['prediction_label']
churned.value_counts()

prediction_label
retained    4154
churned      136
Name: count, dtype: int64

* The model predicts only 136 churn while it is actually 761 customer churn
* Applying feature engineering might improve the accuracy
* It is also due to feature imbalace , rescaling the data could help improve the accuracy

In [74]:
# Test the model using test data
df_test=df_test.dropna(axis=0)
test =predict_model(gbst, data=df_test)
test.head()

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
0   1806         0       0       37.168690                     3098   
1   3451         0       0      210.090088                     2873   
2   6375         0       0       99.767601                     1386   
3   8122         0       0      132.322357                       89   
4  10586         0       0      222.107437                      980   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     236                       0       5570.745605   
1                     210                      35      12012.320312   
2                      21                       0       4169.721680   
3                     174                     104       5044.894043   
4                       0                       6       4932.029297   

   duration_minutes_drives  activity_days  driving_days   device  type  \
0               943.950134              6             0  Android  test   
1              4804.667480              1             0  Android  test   
2              3154.855957              2             0  Android  test   
3              2205.093994              4             0   iPhone  test   
4              2656.000000              3             0  Android  test   

      label prediction_label  prediction_score  
0  retained         retained            0.8072  
1  retained         retained            0.7543  
2  retained         retained            0.5627  
3  retained         retained            0.5533  
4   churned         retained            0.5184

In [72]:
#actual
churned= test['label']
churned.value_counts()

label
retained    67
churned     30
Name: count, dtype: int64

In [75]:
# model output
churned= test['prediction_label']
churned.value_counts()

prediction_label
retained    88
churned      9
Name: count, dtype: int64